### DISCUSSION TITLE
Auto references with JSON artifact type

### DISCUSSION THREAD

**Discussion by lekhamaru** [Apicurio/apicurio-registry#5584] [Q&A]
I'm using the latest version of server 3.0.3.
While using the maven plugin to auto register references, references attached to "allOf" do not get referenced while registering the main artifact.
is that expected behavior?

**Comment by Ganesh-403** [comment]
@lekhamaru Based on the current documentation, I would expect **`autoRefs` to detect all artifact references that the plugin knows how to discover**, regardless of whether they appear under `properties`, `allOf`, `oneOf`, etc. There isn't any documented limitation stating that references inside `allOf` are intentionally ignored.

The documentation simply describes `autoRefs` as:

> Automatically detect and register all references to artifacts in the same directory.

It doesn't call out any exclusions for JSON Schema constructs such as `allOf`. Additionally, JSON Schema is listed as one of the artifact types that supports automatic reference detection. :contentReference[oaicite:0]{index=0}

So, if your schema looks something like:

```json
{
  "allOf": [
    { "$ref": "common.json" },
    {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        }
      }
    }
  ]
}
```

I would expect `common.json` to be detected and registered when:

```xml
<autoRefs>true</autoRefs>
```

is enabled and both schemas are located in the same directory.

From what I can tell, I couldn't find any documentation confirming that references under `allOf` are intentionally skipped. That makes me suspect one of two possibilities:

1. The current implementation of the auto-reference scanner doesn't traverse `allOf` (which would be a bug or missing feature).
2. There's a limitation in the JSON Schema reference detection logic that simply hasn't been documented yet.

As a workaround, you can explicitly declare the references instead of relying on `autoRefs`, for example:

```xml
<references>
    <reference>
        <groupId>test-group</groupId>
        <artifactId>common</artifactId>
        <version>1</version>
        <artifactType>JSON</artifactType>
        <file>${project.basedir}/src/main/resources/common.json</file>
    </reference>
</references>
```

This should register the dependency regardless of where the `$ref` appears in the schema. :contentReference[oaicite:1]{index=1}

Could you share a minimal example of the parent schema containing the `allOf` and the referenced schema? It would also be helpful to know whether the `$ref` is:

- directly inside the `allOf` array,
- nested inside an element of `allOf`,
- or combined with other keywords.

That would help determine whether this is an unsupported edge case or a bug in the `autoRefs` implementation.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
